File: src/leetcode/java/ConstructStringFromBinaryTree_606.java
package leetcode.java;

public class ConstructStringFromBinaryTree_606 {
  public String tree2str(TreeNode t) {
    if (t == null) return "";
    StringBuilder b = new StringBuilder();
    appendNode(t, b);
    return b.toString();
  }

  private void appendNode(TreeNode t, StringBuilder b) {
    b.append(t.val);
    if (t.left != null) preTrav(t.left, b);
    if (t.right != null) {
      if (t.left == null) {
        b.append("()");
      }
      preTrav(t.right, b);
    }
  }

  private void preTrav(TreeNode t, StringBuilder b) {
    b.append("(");
    appendNode(t, b);
    b.append(")");
  }

  class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
      val = x;
    }
  }
}

File: src/leetcode/python/ConstructStringFromBinaryTree_606.py
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class ConstructStringFromBinaryTree_606(object):
    def tree2str(self, t):
        """
        :type t: TreeNode
        :rtype: str
        """
        if t is None:
            return ""
        res = []

        def preTrav(node):
            res.append('(')
            appendNode(node)
            res.append(')')

        def appendNode(treenode):
            res.append(str(treenode.val))
            if treenode.left:
                preTrav(treenode.left)
            if treenode.right:
                if treenode.left is None:
                    res.append("()")
                preTrav(treenode.right)

        appendNode(t)
        return "".join(str(res))
